servers/visual_server.cpp

```diff
@@ -768,6 +769,14 @@ Error VisualServer::_surface_set_data(Array p_arrays, uint32_t p_format, uint32_
 				ERR_FAIL_COND_V(indices.size() == 0, ERR_INVALID_PARAMETER);
 				ERR_FAIL_COND_V(indices.size() != p_index_array_len, ERR_INVALID_PARAMETER);
 
+				// Vertex cache optimization?
+				if (p_format & ARRAY_FLAG_USE_VERTEX_CACHE_OPTIMIZATION) {
+					// Expecting triangles.
+					ERR_FAIL_COND_V((indices.size() % 3) != 0, ERR_INVALID_PARAMETER);
+					VertexCacheOptimizer opt;
+					opt.reorder_indices_pool(indices, indices.size() / 3, p_vertex_array_len);
```

Comment:
Would it be possible to do this earlier in the pipeline?

One side effect of doing it here is that the CPU and GPU representations of the mesh will get out of sync. For generated meshes, that would happen regardless, but for imported meshes, it might be an issue. When reading back a mesh from the RenderingServer, it will have these changes, but the version the user sent to the RenderingServer won't

That being said, unless something changed recently, we used to do a round trip to the GPU during mesh import, which means that this code will run during import anyway. The only problem is that it would then needlessly run again at run time

Reply:
Good point, will have a look tomorrow.

Given this is controlled by a flag `ARRAY_FLAG_USE_VERTEX_CACHE_OPTIMIZATION` it might be possible to do something like remove the flag when loading from a file (rather than the import, or creating geometry at runtime).

Reply:
Ok reminding myself of how this works, the flag `ARRAY_FLAG_USE_VERTEX_CACHE_OPTIMIZATION` is optional (it is _not_ part of `ARRAY_COMPRESS_DEFAULT`) and thus is only currently called during the import.

So due to the round trip on importing, the data saved after import _is_ vertex cache optimized, but the optimization isn't run again at runtime while loading the files (I've just tested and confirmed this).

This does mean for user generated content, the user would need to explicitly set the flag if they want vertex cache optimization, but this seems reasonable as it may not be wanted in all cases (and particularly not for dynamic content).

I seem to remember considering these problems when writing the PR and stepped back from making it "fully automatic in all cases", to having to explicitly choose the optimization flag, to avoid applying it multiple times as you point out.

Reply:
I see. That makes sense!